lib/services/address.js

```diff
@@ -0,0 +1,131 @@
+'use strict';
+
+var Promise = require('bluebird');
+var bitcore = require('bitcore');
+var TransactionService = require('./transaction');
+var _ = bitcore.deps._;
+
+var NULLTXHASH = bitcore.util.buffer.emptyBuffer(32).toString('hex');
+var LASTTXHASH = bitcore.util.buffer.fill(bitcore.util.buffer.emptyBuffer(32), -1).toString('hex');
+var MAXOUTPUT = 4294967295;
+
+function AddressService(opts) {
+  opts = _.extend({}, opts);
+  this.transactionService = opts.transactionService;
+  this.blockService = opts.blockService;
+  this.database = opts.database || Promise.promisifyAll(new LevelUp(config.get('LevelUp')));
+  this.rpc = opts.rpc || Promise.promisifyAll(new RPC(config.get('RPC')));
+}
+
+AddressService.prototype.getSummary = function(address, confirmations) {
+
+  var self = this;
+  var tip, allOutputs, spent;
+  confirmations = confirmations || 6;
+
+  return Promise.try(function() {
+
+    return self.blockService.getLatest();
+
+  }).then(function(latest) {
+
+    tip = latest;
+    return self.getAllOutputs(address);
+
+  }).then(function(outputs) {
+
+    allOutputs = outputs;
+    return self.getSpent(address);
+
+  }).then(function(spent) {
+
+    return self.buildAddressSummary(address, tip, allOutputs, spent, confirmations);
+
+  });
+};
+
+AddressService.prototype.getAllOutputs = function(address) {
+  var results = [];
+  var self = this;
+
+  return new Promise(function(resolve, reject) {
+    self.database.createReadStream({
+      gte: TransactionService.Index.getOutputsForAddress(address, NULLTXHASH, 0),
+      lte: TransactionService.Index.getOutputsForAddress(address, LASTTXHASH, MAXOUTPUT)
+    }).on('data', function(element) {
+      results.push(element.value);
+    }).on('close', function() {
+      reject();
+    }).on('end', function() {
+      resolve(results);
```

Comment:
There may be a race condition between `results.push()` and `resolve(results)`. If the 'data' and 'end' events happen at nearly the same time, results may not have the pushed item by the time `resolve` is called. May not be an issue since it's only using `push`, I've had an issue when data handler was doing more, and was fixed by piping to a writable stream using [callback-stream](https://www.npmjs.com/package/callback-stream)